### PR TITLE
Fix deprecated uses of Homebrew args methods

### PR DESCRIPTION
**Pull request by SeekingMeaning**
- [ ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Related: #8144

Warning: Not thoroughly tested

Tested with:

- `brew deps --include-requirements` at commit 0fdf8f2 before it was fixed
- `brew livecheck` at commit https://github.com/Homebrew/homebrew-livecheck/commit/aafb126ada9c0625929aa4406532021a83ca8a4b before it was fixed
- [`beeftornado/rmtree`](https://github.com/beeftornado/homebrew-rmtree)
- [`claui/whence`](https://github.com/claui/homebrew-whence)

Todo:

- [x] Reduce warning spam
  - Example: https://gist.github.com/SeekingMeaning/4ee8dc139dbcd28c8d8c6f8e41dea188
- [x] Move now-deprecated `Homebrew.argv_includes_ignores` to `compat/dependencies_helpers.rb`

### CODE REVIEW

**Comment by reitermarkus** [comment]
I still stand by my point that this is an internal API. The fact that some external commands use private APIs does not make them public. Otherwise we will need to treat every API as public since it may be used by some external command and deprecate every single method.

This shouldn't be blocker however, given the amount of breakage this has caused.

File: Library/Homebrew/global.rb [hunk 1]
```diff
@@ -82,7 +84,12 @@ def failed?
     end
 
     def args
-      @args ||= CLI::Args.new
+      @args ||= CLI::Args.new(global: true)
+      unless @args.global || @printed_args_warning
```
Comment (MikeMcQuaid) [inline Library/Homebrew/global.rb:88]: Can this be reworked to be an `if` or an early return and multiple conditions? `unless ||` is hard to follow. Thanks!

**Comment by MikeMcQuaid** [comment]
> The fact that some external commands use private APIs does not make them public.

I disagree. You need(ed) to use these APIs in order to get correct `--help`  output and argument parsing for external commands. We should avoid breaking external commands when we can avoid it and this seems like a good compromise I'd like to land before we make a new release.

**Comment by reitermarkus** [comment]
> You need(ed) to use these APIs in order to get correct `--help`  output

Is this documented anywhere? The docs explicitly link to https://github.com/Homebrew/homebrew-services/blob/a58a1fe9145de4e50e1cbfb5b0e8a30087826393/cmd/brew-services.rb#L1-L23.

**Comment by MikeMcQuaid** [comment]
> Is this documented anywhere?

No, it seems not.

**Comment by reitermarkus** [comment]
> No, it seems not.

Hence why I consider is a private API.

**Comment by MikeMcQuaid** [comment]
@reitermarkus You can consider it to be or not to be all you want. I don't actually think it matters if it is or not: if we can avoid breaking other people's code then it seems considerate to do so.

**Comment by reitermarkus** [comment]
> You can consider it to be or not to be all you want.

I think we can agree that it *should* be public.

File: Library/Homebrew/global.rb (hunk 1 above)
Comment (reitermarkus) [reply Library/Homebrew/global.rb:88]: I'm not sure when `global` will ever be `true` since `Homebrew.args` is assigned in `brew.rb` with `global: false`.

**Comment by SeekingMeaning** [comment]
Question: Will `Homebrew.args` be completely deprecated or only certain uses of it? i.e. will checks for global options like `Homebrew.args.verbose?` be deprecated too?

**Comment by reitermarkus** [comment]
> Maybe all calls to `Homebrew.args` should print the deprecation warning, or at least temporarily

`debug?`, `quiet?`, `verbose?` should not show a warning.

**Comment by reitermarkus** [comment]
In https://github.com/Homebrew/brew/pull/8187 i introduced a `Context` module to replace the remaining global `Homebrew.args`.

**Comment by reitermarkus** [comment]
I think we need some sort of `DeprecatedArgs` class:


```ruby
class DeprecatedArgs < Args
  def respond_to_missing?(*)
    super
  end

  def method_missing(method, *)
    unless [:debug?, :quiet?, :verbose?].include?(method)
      odeprecated ...
    end
  
    super
  end
end
```

**Comment by Bo98** [comment]
I consider anything that appears on https://rubydoc.brew.sh as "public API".

If it's not public API it should be marked with `# @private` (or maybe `@api private` but we don't seem to be using that), or ideally actually private in Ruby, if possible.

**Comment by reitermarkus** [comment]
> I consider anything that appears on https://rubydoc.brew.sh as "public API".

That would mean everything is public since that is the default. Ideally only things which are actually documented should show up in RubyDoc.

**Comment by Bo98** [comment]
My point is: we point people to RubyDoc. We shouldn't be telling them: "some things are private and aren't indicated as such - good luck!"

It seems to be then we don't have any real indicators of what is public API or not. If `Homebrew.args` is supposed to be private API, we've not done a good job at showing that considering several taps use it and I haven't seen any use the return value of `parse` until recently.

"Documentation" isn't even that indicative currently. [We document things that are denoted as private](https://rubydoc.brew.sh/Language/Python/Virtualenv.html#needs_python%3F-instance_method) and [we don't document things that are definitely public API](https://rubydoc.brew.sh/Formula.html#opt_bin-instance_method).

**Comment by reitermarkus** [comment]
From https://rubydoc.brew.sh/:

> The main class you should look at is the Formula class (and classes linked from there). That’s the class that’s used to create Homebrew formulae (i.e. package descriptions). Assume anything else you stumble upon is private.

**Comment by Bo98** [comment]
So is all of `Homebrew::CLI` private?

**Comment by reitermarkus** [comment]
> So is all of `Homebrew::CLI` private?

Based on the (non-existent) documentation for it, I do consider it private. Whether is *should*  be private, see above.

**Comment by Bo98** [comment]
We should ideally be making an example and stop using private API in our taps.

**Comment by reitermarkus** [comment]
I think our official external commands should be allowed to use internal APIs, but we should be testing them on `brew` changes.
